### ISSUE TITLE
v0.40.1: live runtime weighted-loss combination for preference_loss_weights

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#68]
**Background.** v0.40.0 ships ``training.preference_loss_weights`` as a config-level surface only (validated dict of 2–5 entries summing to 1.0). At runtime, ``PreferenceTrainerWrapper.setup`` raises ``NotImplementedError`` because TRL's preference trainers (DPO/SimPO/ORPO/IPO/BCO) don't expose a clean ``compute_loss`` override hook for blending.

This mirrors the project's repeated stub-then-live pattern: v0.27.0 MII, v0.37.0 multipack, v0.38.0 quant menu, v0.39.0 ReLoRA all shipped schema first then wired live in the ``X.Y.1`` patch.

**Proposed fix path.**

1. Subclass each TRL trainer (``DPOTrainer``, ``CPOTrainer`` for SimPO, ``ORPOTrainer``, ``IPOTrainer``, ``BCOTrainer``) with a single mixin that overrides ``compute_loss``:
   - Compute each component loss in turn (skipping any whose weight isn't in the config).
   - Return the weighted sum: ``loss = Σ (w_i · loss_i)``.
2. Add a runtime dispatcher in ``PreferenceTrainerWrapper.setup`` that picks the right base trainer (the most expensive component, or the first one alphabetically) and wraps it with the mixin.
3. Surface per-component losses in the Rich dashboard (one row per component) so users can see which side is dominating.
4. Remove the ``NotImplementedError`` in ``PreferenceTrainerWrapper.setup`` and replace the ``is_multi_objective_preference`` runtime gate with the live path.

**Acceptance criteria.**

- [ ] ``PreferenceTrainerWrapper.setup`` accepts ``preference_loss_weights`` and wires the mixin without raising.
- [ ] Smoke test: 5-step training run on toy GPT-2 with ``{dpo: 0.7, bco: 0.3}`` completes; final loss is sane (not NaN, not the constant 0).
- [ ] Per-component losses appear in ``trainer.state.log_history`` for inspection.
- [ ] No regressions in existing preference tests (DPO/SimPO/ORPO/IPO/BCO single-loss paths unchanged).
- [ ] Update ``v0.40.0`` Known Limitations entry in CLAUDE.md and SECURITY.md.

**References.**

- v0.40.0 Part D schema validator: ``soup_cli/config/schema.py::_validate_preference_loss_weights``
- Stub-then-live precedent: #65 (multipack v0.37.1), v0.38.0 quant menu, v0.39.0 ReLoRA

Milestone: v0.40.1.

**Comment by MakazhanAlpamys** [comment]
Phase 1 shipped in v0.40.1: `PreferenceTrainerWrapper.setup` no longer raises `NotImplementedError` when `preference_loss_weights` is set. The wrapper now builds the highest-weighted loss as the primary inner trainer and prints the active blend for confirmation. BCO mixed with paired losses is rejected at runtime (data-format incompatible).

The pure-function math kernel (`utils/preference_combine.py` — DPO / IPO / SimPO / ORPO closed-form losses + `combine_losses` weighted sum) is fully shipped and exercised by 24 tests including gradient propagation and NaN propagation.

**Still pending (this issue stays open for v0.40.2):** full per-batch weighted-loss combination across all named losses via TRL trainer subclassing — requires overriding `compute_loss` on `DPOTrainer` / `CPOTrainer` / `BCOTrainer` to compute each loss formula on the same forward pass and combine via `combine_losses`. Mechanical per-trainer expansion.

Try the primary-loss form now:
```yaml
task: preference
training:
  preference_loss_weights:
    dpo: 0.7
    simpo: 0.3
```

See [release notes](https://github.com/MakazhanAlpamys/Soup/releases/tag/v0.40.1).

**Comment by MakazhanAlpamys** [comment]
Shipped in v0.53.11. See [release notes](https://github.com/MakazhanAlpamys/Soup/releases/tag/v0.53.11).

Try: `task: preference` + `preference_loss_weights: {dpo: 0.6, simpo: 0.4}` in your soup.yaml `training:` block.
